### PR TITLE
Upgrade anyio to v3

### LINKED ISSUE #487: Anyio version 3 breaks AsyncContentsManager
## Description

Anyio version 3 was recently introduced and we found that the various AsyncContentsManager calls that use anyio's `run_sync_in_worker_thread` method fail with a traceback similar to:
```
HTTPServerRequest(protocol='http', host='127.0.0.1:38403', method='GET', uri='/a%40b/api/contents/Directory%20with%20spaces%20in', version='HTTP/1.1', remote_ip='127.0.0.1')
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/anyio/_backends/_asyncio.py", line 697, in run_sync_in_worker_thread
    work_queue = _threadpool_work_queue.get()
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/anyio/lowlevel.py", line 120, in get
    raise LookupError(f'Run variable "{self._name}" has no value and no default set')
LookupError: Run variable "_threadpool_work_queue" has no value and no default set

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/tornado/web.py", line 1704, in _execute
    result = await result
  File "/home/runner/work/jupyter_server/jupyter_server/jupyter_server/services/contents/handlers.py", line 111, in get
    path=path, type=type, format=format, content=content,
  File "/home/runner/work/jupyter_server/jupyter_server/jupyter_server/utils.py", line 176, in ensure_async
    result = await obj
  File "/home/runner/work/jupyter_server/jupyter_server/jupyter_server/services/contents/filemanager.py", line 692, in get
    model = await self._dir_model(path, content=content)
  File "/home/runner/work/jupyter_server/jupyter_server/jupyter_server/services/contents/filemanager.py", line 574, in _dir_model
    dir_contents = await run_sync_in_worker_thread(os.listdir, os_dir)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/anyio/to_thread.py", line 37, in run_sync_in_worker_thread
    return await run_sync(func, *args, cancellable=cancellable, limiter=limiter)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/anyio/to_thread.py", line 29, in run_sync
    limiter=limiter)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/anyio/_backends/_asyncio.py", line 707, in run_sync_in_worker_thread
    find_root_task().add_done_callback(_loop_shutdown_callback)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/anyio/_backends/_asyncio.py", line 56, in find_root_task
Error:     raise RuntimeError('Cannot find root task for setting cleanup callback')
```
The [failure was also seen on a build of PR 483](https://github.com/jupyter-server/jupyter_server/runs/2399817573?check_suite_focus=true#step:9:600) prior to [capping `anyio < 3`](https://github.com/jupyter-server/jupyter_server/commit/2b51ee37bdad305cb349e246c8ba94381cdb2048).

## Reproduce

1. From the jupyter_server repo
2. Ensure `anyio` version 3 is installed (released on April 20)
3. Run `pytest -v jupyter_server/tests/services/contents/test_largefilemanager.py::test_save`
4. See failure similar to above.

## Expected behavior

Tests should pass.

## Context

- Operating System and version: MacOS/Ubuntu (not sure about Windows)
- Browser and version: N/A
- Jupyter Server version: master

I noticed the following deprecation warning when running `pytest`:
```
jupyter_server/tests/services/contents/test_largefilemanager.py::test_save[AsyncLargeFileManager]
  /opt/anaconda3/envs/jupyter-server-dev/lib/python3.7/site-packages/anyio/to_thread.py:36: DeprecationWarning: run_sync_in_worker_thread() has been deprecated, use anyio.to_thread.run_sync() instead
    DeprecationWarning)
```
However, a quick attempt to switch to `anyio.to_thread.run_sync()` resulted in the same issue since it appears `run_sync()` still calls `run_sync_in_worker_thread()` internally.  As a result, this may be an `anyio` issue.

<details><summary>Using run_sync()</summary>
<pre>
ERROR    traitlets:filemanager.py:746 Error while saving file: Untitled.ipynb [Errno 2] No such file or directory: '/private/var/folders/4w/qxgsbhyx1y93qlk9mnzwjryc0000gn/T/pytest-of-kbates/pytest-74/test_save_AsyncLargeFileManage0/.~Untitled.ipynb' -> '/private/var/folders/4w/qxgsbhyx1y93qlk9mnzwjryc0000gn/T/pytest-of-kbates/pytest-74/test_save_AsyncLargeFileManage0/Untitled.ipynb'
Traceback (most recent call last):
  File "/opt/anaconda3/envs/jupyter-server-dev/lib/python3.7/site-packages/anyio/_backends/_asyncio.py", line 697, in run_sync_in_worker_thread
    work_queue = _threadpool_work_queue.get()
  File "/opt/anaconda3/envs/jupyter-server-dev/lib/python3.7/site-packages/anyio/lowlevel.py", line 120, in get
    raise LookupError(f'Run variable "{self._name}" has no value and no default set')
LookupError: Run variable "_threadpool_work_queue" has no value and no default set

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/kbates/repos/oss/jupyter/jupyter_server/jupyter_server/services/contents/fileio.py", line 121, in atomic_writing
    yield fileobj
  File "/Users/kbates/repos/oss/jupyter/jupyter_server/jupyter_server/services/contents/fileio.py", line 218, in atomic_writing
    yield f
  File "/Users/kbates/repos/oss/jupyter/jupyter_server/jupyter_server/services/contents/fileio.py", line 389, in _save_notebook
    await run_sync(partial(nbformat.write, version=nbformat.NO_CONVERT), nb, f)
  File "/opt/anaconda3/envs/jupyter-server-dev/lib/python3.7/site-packages/anyio/to_thread.py", line 29, in run_sync
    limiter=limiter)
  File "/opt/anaconda3/envs/jupyter-server-dev/lib/python3.7/site-packages/anyio/_backends/_asyncio.py", line 707, in run_sync_in_worker_thread
    find_root_task().add_done_callback(_loop_shutdown_callback)
  File "/opt/anaconda3/envs/jupyter-server-dev/lib/python3.7/site-packages/anyio/_backends/_asyncio.py", line 56, in find_root_task
    raise RuntimeError('Cannot find root task for setting cleanup callback')
RuntimeError: Cannot find root task for setting cleanup callback

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/kbates/repos/oss/jupyter/jupyter_server/jupyter_server/services/contents/filemanager.py", line 732, in save
    await self._save_notebook(os_path, nb)
  File "/Users/kbates/repos/oss/jupyter/jupyter_server/jupyter_server/services/contents/fileio.py", line 389, in _save_notebook
    await run_sync(partial(nbformat.write, version=nbformat.NO_CONVERT), nb, f)
  File "/opt/anaconda3/envs/jupyter-server-dev/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/Users/kbates/repos/oss/jupyter/jupyter_server/jupyter_server/services/contents/fileio.py", line 218, in atomic_writing
    yield f
  File "/opt/anaconda3/envs/jupyter-server-dev/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/Users/kbates/repos/oss/jupyter/jupyter_server/jupyter_server/services/contents/fileio.py", line 125, in atomic_writing
    replace_file(tmp_path, path)
  File "/Users/kbates/repos/oss/jupyter/jupyter_server/jupyter_server/services/contents/fileio.py", line 35, in replace_file
    os.replace(src, dst)
FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/4w/qxgsbhyx1y93qlk9mnzwjryc0000gn/T/pytest-of-kbates/pytest-74/test_save_AsyncLargeFileManage0/.~Untitled.ipynb' -> '/private/var/folders/4w/qxgsbhyx1y93qlk9mnzwjryc0000gn/T/pytest-of-kbates/pytest-74/test_save_AsyncLargeFileManage0/Untitled.ipynb'
</pre>
</details>

cc: @mwakaba2

### PR DESCRIPTION
**Pull request by mwakaba2**
This fixes #487 (Anyio version 3 breaks AsyncContentsManager).

### Changes

- Upgrade anyio from v2 to v3.
- Replace deprecated function `run_sync_in_worker_thread` with `to_thread.run_sync`.
- Tests: Prefix contents manager fixture with `jp`

### How to test
```bash
$  pip install -e ".[test]"
$  pip install git+https://github.com/agronholm/anyio.git@fix-264
$  python -m pytest -v jupyter_server/tests/services/contents/
```

This PR depends on Anyio's [Refactored asyncio thread pooling](https://github.com/agronholm/anyio/pull/268) .

### Todo

- [x] Test against published anyio version that includes the [Refactored asyncio thread pooling](https://github.com/agronholm/anyio/pull/268) PR.

### CODE REVIEW

File: setup.cfg
Comment (mwakaba2) [inline setup.cfg:45]: TODO: Update the lower bound version to the new version with the fix.

Comment (agronholm) [reply setup.cfg:45]: Also the syntax is incorrect (missing the comma).

Comment (mwakaba2) [reply setup.cfg]: Ah, good catch. Thanks! I fixed it.

**Comment by kevin-bates** [comment]
Excellent, thank you @mwakaba2 and @agronholm!  I'm going to place this into DRAFT for now until Alex's anyio PR is merged and released.

**Comment by kevin-bates** [comment]
Good to see the quick turnaround on the anyio patch, although it looks like 3.6 still has some heartburn.  

Choices...
1. Cap anyio < 3 for python < 3.7 and alias `run_sync` to `run_sync_in_worker_thread` on 3.6?
2. Disable `AsyncContentsManager` on 3.6 given it is EOL at EOY and probably has few consumers at this point?
3. Try to determine what it is about 3.6 that's causing the hang?
4. ?

The PR for #490 will probably need option 1.

**Comment by yan12125** [comment]
> Disable AsyncContentsManager on 3.6 given it is EOL at EOY and probably has few consumers at this point?

FWIW, according to https://pypistats.org/packages/jupyter-server, Python 3.6 downloads account for 13%~36% on PyPI for the past 30 days.

**Comment by kevin-bates** [comment]
Thanks for the metric @yan12125 - that's good to know, although I suspect the percentage overriding the default contents manager to the async version is nearly zero.  Nevertheless, I think the first option would be the best step forward (barring a more graceful solution) as it can also be used to accomplish #490.

**Comment by mwakaba2** [comment]
Yeah I think the first option sounds good to unblock #490. I'll also look into the issue with python 3.6 and check with anyio if it's related to their new update.

**Comment by agronholm** [comment]
If the problem with py3.6 is caused by AnyIO, I would like to know.

**Comment by mwakaba2** [comment]
@kevin-bates I'd appreciate your review on this PR. 

@agronholm I created a new issue for the py3.6 error. #505 It looks like it's an anyio issue. 
I'd appreciate if you can take a look. I'll be able to debug further after this Wednesday.

**Review by kevin-bates — APPROVED** [review]
These changes look good and allow us to proceed with #490.  Thank you!

**Comment by mwakaba2** [comment]
thank you @kevin-bates ! I don't have permission to merge the pull request, so if it all looks good, please feel free to merge it.

**Comment by kevin-bates** [comment]
Thanks Mariko - I was giving others a chance to comment, but these seem straightforward enough - so I'll go ahead.

**Comment by Zsailer** [comment]
@mwakaba2, I've upgraded your permissions to "write" 😎 . My apologies for not doing this sooner!

**Comment by mwakaba2** [comment]
sweet! thanks so much. I look forward to contributing more! 😎

**Pull request merged**
